### PR TITLE
Add config agent core ngt test

### PR DESCRIPTION
**Pull request by hlts2**
### Description:

I added test case for `pkg/agent/core/ngt/config`.
and I refactored.

Why:
Only GlobalConfig was bound and binding to other elements failed.(`Server`, `Observability`, `NGT`)
Because when we unmarsh a json in a struct, if this struct has `UnmarshalJSON`, golang will call that method and exit json's unmarshal. Since GlobalConfig is an embedded structure this time, it was judged that the method was defined in the root, and the process was completed.

What:
I deleted `UnmarshalJSON` method of GlobalConfig because we can bind without that method.

### Environment:

- Go Version: 1.16
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [x] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/agent/core/ngt/config/config_test.go [hunk 1]
```diff
@@ -52,31 +56,278 @@ func TestNewConfig(t *testing.T) {
...
+		func() test {
+			data := `
+version: v1.0.0
+server_config:
+  full_shutdown_duration: 10ms
+observability:
+  enabled: true
+ngt:
+  index_path: /var/index
+`
```
[the remark this answers is not in the document]
Comment (hlts2) [reply pkg/agent/core/ngt/config/config_test.go]: thank you.
fixed: https://github.com/vdaas/vald/pull/1219/commits/d7ef9f549e01bf7e8eed595a03802d1981794471

File: internal/config/config.go [hunk 2]
```diff
@@ -50,6 +50,9 @@ const (
 
 // Bind binds the actual data from the receiver field.
 func (c *GlobalConfig) Bind() *GlobalConfig {
+	if c == nil {
+		c = new(GlobalConfig)
+	}
```
Comment (vankichi) [inline internal/config/config.go:55]: you should create the test for covering these codes.
I'm not sure this code is correct?
I think when `c==nil`, `Bind()` method cannot be running 🤔

File: pkg/agent/core/ngt/config/config.go [hunk 3]
```diff
@@ -46,7 +48,7 @@ func NewConfig(path string) (cfg *Data, err error) {
 	}
 
 	if cfg != nil {
-		cfg.Bind()
```
Comment (kevindiu) [inline pkg/agent/core/ngt/config/config.go:49]: Can  you tell me why do you remove the cfg.Bind() here?

File: pkg/agent/core/ngt/config/config.go [hunk 4]
```diff
@@ -46,7 +48,7 @@ func NewConfig(path string) (cfg *Data, err error) {
...
+		cfg.GlobalConfig.Bind()
```
Comment (kevindiu) [inline pkg/agent/core/ngt/config/config.go:51]: I think cfg.GlobalConfig nil check is necessary

File: internal/config/config.go (hunk 2 above)
Comment (hlts2) [reply internal/config/config.go:55]: thank you.

> I think when c==nil, Bind() method cannot be running 🤔
got it, I think so.
I think we should delete it

File: pkg/agent/core/ngt/config/config.go (hunk 3 above)
Comment (hlts2) [reply pkg/agent/core/ngt/config/config.go:49]: sorry. this is my misstake.
I think `cfg.Bind()` is correct :bow:

File: pkg/agent/core/ngt/config/config.go (hunk 4 above)
Comment (hlts2) [reply pkg/agent/core/ngt/config/config.go:51]: thank you for comment.

`GlobalConfig` is not pointer type in current code.
so I think we need not check nil.

File: pkg/agent/core/ngt/config/config.go (hunk 3 above)
Comment (hlts2) [reply pkg/agent/core/ngt/config/config.go]: fixed: https://github.com/vdaas/vald/pull/1219/commits/164baadf6da88821fcb5eed9196dae2cd0c11a77

File: internal/config/config.go (hunk 2 above)
Comment (hlts2) [reply internal/config/config.go]: fixed: https://github.com/vdaas/vald/pull/1219/commits/760d84928e84d06bc026398dffcce0c088d7e49c

File: pkg/agent/core/ngt/config/config_test.go [hunk 5]
```diff
@@ -52,31 +56,278 @@ func TestNewConfig(t *testing.T) {
...
+					wantCfg: nil,
+					err: &fs.PathError{
+						Op:   "open",
+						Path: path,
+						Err:  syscall.Errno(0x2),
+					},
+				},
+			}
+		}(),
 	}
```
Comment (kevindiu) [inline pkg/agent/core/ngt/config/config_test.go:331]: can you add more case? like empty config file, no read permission of config file.

File: pkg/agent/core/ngt/config/config_test.go [hunk 6]
```diff
@@ -52,31 +56,278 @@ func TestNewConfig(t *testing.T) {
...
+							VQueue:    new(config.VQueue),
+						},
+					},
+					err: nil,
+				},
+			}
+		}(),
+		func() test {
+			data := `
+                version: v1.0.0
```
[the remark this answers is not in the document]
Comment (hlts2) [reply pkg/agent/core/ngt/config/config_test.go:191]: thank you.
I fixed: https://github.com/vdaas/vald/pull/1219/commits/0f43ebe9210893b1d881af774297e3c35c55140c

File: pkg/agent/core/ngt/config/config_test.go [hunk 7]
```diff
@@ -52,31 +56,278 @@ func TestNewConfig(t *testing.T) {
...
+						},
+					},
+					err: nil,
+				},
+			}
+		}(),
+		func() test {
+			path := "not_found.txt"
+			return test{
+				name: "test_case_2",
```
[the remark this answers is not in the document]
Comment (hlts2) [reply pkg/agent/core/ngt/config/config_test.go]: thank you.
I fixed: 0f43ebe

File: pkg/agent/core/ngt/config/config_test.go (hunk 5 above)
Comment (hlts2) [reply pkg/agent/core/ngt/config/config_test.go:423]: thank you
I fixed: https://github.com/vdaas/vald/pull/1219/commits/c3fcd2c20cf3ac721a26fc79abe9df5f6ff7380c

**Review by kevindiu — DISMISSED** [review]
LGTM

**Comment by kevindiu** [comment]
I think this PR requires author review.

**Review by vankichi — DISMISSED** [review]

**Comment by kevindiu** [comment]
The CI error occurred, please check it.
https://github.com/vdaas/vald/pull/1219/checks?check_run_id=2596681845

**Review by kpango — APPROVED** [review]
LGTM

**Comment by kpango** [comment]
seems pkg test failed can you fix it??

and if pkg test passed can you test e2e-deploy and e2e-chaos??

**Comment by hlts2** [comment]
@kpango 
> seems pkg test failed can you fix it??
and if pkg test passed can you test e2e-deploy and e2e-chaos??

thank you for review.
I got it. I will do it!

**Pull request merged**